mesonbuild/modules/fs.py

```diff
@@ -89,6 +89,73 @@ def as_posix(self, state: 'ModuleState', args: T.Sequence[str], kwargs: dict) ->
+        If a third path is given in 'within' argument, then the function will return the first
+        argument unchanged if it is not within the 'within' path.
+        """
+        if len(args) != 2:
+            raise MesonException('fs.relative_to takes two arguments and optionally a "within" and a "native" argument.')
+        # pathlib requires to use PureWindowsPath for Windows paths for absolute and relative
+        # path computations
+        for_machine = self.interpreter.machine_from_native_kwarg(kwargs)
+        if for_machine == MachineChoice.BUILD:
+            system = state.build_machine.system
+        else:
+            system = state.host_machine.system
+        if system == "windows":
+            path_class = PureWindowsPath  # type: T.Union[T.Type[PurePosixPath], T.Type[PureWindowsPath]]
+        else:
+            path_class = PurePosixPath
+        
+        path_to = path_class(args[0])
+        if not path_to.is_absolute():
+            raise MesonException(f'The first argument ({path_to}) must be an absolute path.')
+        path_from = path_class(args[1])
+        if not path_from.is_absolute():
+            raise MesonException(f'The second argument ({path_from}) must be an absolute path.')
+        if "within" in kwargs:
+            path_within = path_class(kwargs["within"])
+            if not path_within.is_absolute():
+                raise MesonException('The "within" argument ({path_within}) must be an absolute path.')
+            # Return path_to if it is not relative to path_within
+            try:
+                path_to.relative_to(path_within)
+            except ValueError:
+                return ModuleReturnValue(str(path_to), [])
+        # If path_to starts with path_from, then .relative_to() provides a solution:
+        try:
+            x = path_to.relative_to(path_from)
+        except ValueError:
+            pass
+        else:
+            return ModuleReturnValue(str(x), [])
+
+        # Get the common root between the two paths:
+        parts_to = path_to.parts
+        parts_from = path_from.parts
+        if parts_to[0] != parts_from[0]:
+            raise MesonException(f"{path_to} and {path_from} do not have a common root." +
+                                 "Use the \"within\" argument if you want an absolute path instead of an error.")
+        parts_root = []
+        for part_to, part_from in zip(parts_to, parts_from):
+            if part_to != part_from:
+                break
+            parts_root.append(part_to)
+        root = path_class(parts_root[0]) / path_class("/".join(parts_root[1:]))
+
+        # Then find how many levels do we need to go from path_from to the root:
+        num_levels = len(path_from.parts) - len(root.parts)
+        # And build the final path, going from path_from to the root and then to path_to:
+        path = path_class("/".join(num_levels*['..'])) / path_to.relative_to(root)
+        return ModuleReturnValue(str(path), [])
```

Comment:
This extremely hairy, scary code seems to be entirely a result of a single devastating design flaw:

`os.path.relpath()` is the one true function for computing relations between paths; meanwhile, pathlib is often opinionatedly wrong and this is one of those cases. The proper situation in which pathlib should be used is as an os.PathLike implementation passed to `os.path.*` functions, and:
- `relative_to()`
- in general any concrete Path() (as opposed to the PurePath() functions which do not provide I/O) method is probably horrifically broken and should be avoided, most particularly .exists() as it may arbitrarily traceback instead of returning False when the answer is "it doesn't exist"

Reply:
I agree, however I was asked to move to `pathlib` on the first review here: https://github.com/mesonbuild/meson/pull/7908#issuecomment-717816900. That led me down a rabbit hole in pathlib reaching a python enhancement pull request yet to be merged... (Also liked in this PR)

To be fully honest, after seeing differences of opinions across meson maintainers without a clear path forward for me to follow and get the feature merged, I paused pushing for this. This has also blocked my adoption of meson on a project where I wanted to use this feature.

I understand everyone's opinions and you all have very valid points. I also appreciate the care and time you all put into your review. I just did not have the time to keep pushing for something that I had no idea if it would ever be merged.

Hopefully there will at some point be a consensus of what the meson interface should be and the requirements for the behaviour in corner cases. Once that's clear, then I --or anyone else interested-- will be able to provide a straightforward implementation.

Thanks to all of you for meson and for all your time. I feel bad for criticising your work when you are all giving it for free to the community, I hope you understand I say this with my best intentions, hoping that it helps you to consider possible management improvements in the future. If I am the only one feeling like this, please ignore me. But please take note if you find more people feeling the same way.

Reply:
I changed it to `os.path.relpath()`